espnet2/enh/espnet_enh_s2t_model.py

```diff
@@ -0,0 +1,256 @@
+
+        # Calculating enhancement loss
+        utt_id = kwargs.get("utt_id", None)
+        bypass_enh_flag, skip_enhloss_flag = False, False
+        if utt_id is not None:
+            # TODO(xkc): to pass category info and use predefined category list
+            if utt_id[0].endswith("SIMU"):
+                # For simulated single-/multi-speaker data
+                # feed it to Enhancement and calculate loss_enh
+                bypass_enh_flag = False
+                skip_enhloss_flag = False
+            elif utt_id[0].endswith("REAL"):
+                # For single-speaker real data
+                # feed it to Enhancement but without calculating loss_enh
+                bypass_enh_flag = False
+                skip_enhloss_flag = True
+            else:
+                # For clean data
+                # feed it to Enhancement, without calculating loss_enh
+                bypass_enh_flag = True
+                skip_enhloss_flag = True
+
+        # Bypass the enhancement module
+        if (
+            self.training and skip_enhloss_flag and not bypass_enh_flag
+        ):  # For single-speaker real data: possibility to bypass frontend
+            if random.random() <= self.bypass_enh_prob:
+                bypass_enh_flag = True
+
+        # 1. Enhancement
+        # model forward
+        loss_enh = None
+        if not bypass_enh_flag:
+            (
+                speech_pre,
+                feature_mix,
+                feature_pre,
+                others,
+            ) = self.enh_model.forward_enhance(speech, speech_lengths)
+            # loss computation
+            if not skip_enhloss_flag and self.enh_loss_weight > 0:
+                loss_enh, _, _ = self.enh_model.forward_loss(
+                    speech_pre,
+                    speech_lengths,
+                    feature_mix,
+                    feature_pre,
+                    others,
+                    speech_ref,
+                )
+                loss_enh = loss_enh[0]
+        else:
+            speech_pre = [speech]
+
+        # for data-parallel
+        text = text[:, : text_lengths.max()]
+        if src_text is not None:
+            src_text = src_text[:, : src_text_lengths.max()]
+
+        # 2. ASR or ST
+        if isinstance(self.s2t_model, ESPnetASRModel):  # ASR
+            loss_asr, stats, weight = self.s2t_model(
+                speech_pre[0], speech_lengths, text, text_lengths
+            )
+        elif isinstance(self.s2t_model, ESPnetSTModel):  # ST
+            loss_asr, stats, weight = self.s2t_model(
+                speech_pre[0],
+                speech_lengths,
+                text,
+                text_lengths,
+                src_text,
+                src_text_lengths,
+            )
+        else:
+            raise NotImplementedError(f"{type(self.s2t_model)} is not supported yet.")
+
+        if loss_enh is not None:
+            loss = loss_enh + loss_asr
```

Comment:
We may consider adding weight options for this MTL loss in the future.
e.g., `loss = loss_enh + lambda * loss_asr`

Reply:
Yes. We will update it in the future.